added docker check to workflow

VanshikaSabharwal opened the pull request:
**What kind of change does this PR introduce?**

 bugfix, feature request

**Issue Number:**

Fixes #2402 

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

![image](https://github.com/user-attachments/assets/d649dcfe-f2de-4f6f-b11b-299951f5280d)

**If relevant, did you update the documentation?**

Not relevent

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

No

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced a new job in the GitHub Actions workflow to verify the Talawa Admin application starts correctly in Docker.
  - Added a `Dockerfile` for streamlined application deployment.
  - Included a new section in the installation guide detailing Docker setup, covering prerequisites, development, and production configurations.

- **Enhancements**
  - Updated `.dockerignore` to optimize Docker image builds by excluding unnecessary files.
  - Modified server configuration to change the default port to 4321 and prevent automatic browser opening.

These changes aim to improve the development and deployment experience for users.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,35 @@ jobs:
+                  sudo apt-get update
+                  sudo apt-get install -y docker.io
+        
+              - name: Run Docker Container
+                run: |
+                  docker run -d --name talawa-admin-container -p 8080:8080 talawa-admin-app
+        
```
i did it. 
![image](https://github.com/user-attachments/assets/75a595c3-062f-4acc-be0f-8ee13019c68a)

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,35 @@
       - name: Validate Documents
         run: graphql-inspector validate './src/GraphQl/**/*.ts' './talawa-api/schema.graphql'
 
+  Docker-Start-Check:
+          name: Check if Talawa Admin app starts in Docker
+          runs-on: ubuntu-latest
+          steps:
+             - name: Checkout the Repository
+             uses: actions/checkout@v4
+        
+             - name: Set up Docker
+             run: |
+                  sudo apt-get update
+                  sudo apt-get install -y docker.io
+        
```
did it. 
![Uploading image.png…]()

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,35 @@
+                run: |
+                  docker run -d --name talawa-admin-container -p 8080:8080 talawa-admin-app
+        
+              - name: Check if Talawa Admin App is running
+               run: |
+                nc -z localhost 8080
+        
+                curl --fail --silent http://localhost:8080/index.html || exit 1
+        
```
did  it. 
![Uploading image.png…]()

VanshikaSabharwal replied:
i did it

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,54 @@ jobs:
       - name: Validate Documents
         run: graphql-inspector validate './src/GraphQl/**/*.ts' './talawa-api/schema.graphql'
 
+  Docker-Start-Check:
+          name: Check if Talawa Admin app starts in Docker
+          runs-on: ubuntu-latest
+          needs: [Code-Quality-Checks]
+          steps:
+             - name: Checkout the Repository
+             uses: actions/checkout@v4
+        
+             - name: Set up Docker
+             uses: docker/setup-buildx-action@v3
```
i did it. 
![image](https://github.com/user-attachments/assets/a9d2c472-b735-4699-b942-578f077e2b4c)

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,39 @@
+                run: |
+                  docker run -d --name talawa-admin-app-container -p 8080:8080 talawa-admin-app
+
+            - name: Check if Talawa Admin app starts in Docker
+                runs-on: ubuntu-latest
+                needs: [Code-Quality-Checks]
+                steps:
+                  - name: Checkout the Repository
+                    uses: actions/checkout@v4
```
before i did that only and you told me to do this. you are confusing me. https://github.com/PalisadoesFoundation/talawa-admin/pull/2414

VanshikaSabharwal replied:
yes i did it

varshith257 reviewed (changes requested):
It seems workflow added isn't triggered?

pranshugupta54 reviewed (changes requested):
Looks like there's an error in the YAML file—possibly due to formatting or indentation.

Check: https://github.com/PalisadoesFoundation/talawa-admin/actions/runs/11741762559/workflow#L260

VanshikaSabharwal commented:
> Looks like there's an error in the YAML file—possibly due to formatting or indentation.
> 
> Check: https://github.com/PalisadoesFoundation/talawa-admin/actions/runs/11741762559/workflow#L260

Yes you are right, there is indentation error due to which the workflow is not triggered. I am working on it.

varshith257 commented:
@palisadoes I am still surprised as we don't have any Dockerfile in talawa-admin and how we can support this workflow though 😄

VanshikaSabharwal commented:
> @palisadoes I am still surprised as we don't have any Dockerfile in talawa-admin and how we can support this workflow though 😄

Hey @varshith257, Do i have to create a DockerFile ?

varshith257 commented:
Check #1075

VanshikaSabharwal commented:
> Check #1075

Yes, @varshith257 I checked this and i found out that there still isn't any file. so at this point can i make a docker file?

palisadoes commented:
@VanshikaSabharwal

You'll need to update this PR with:

1. The Dockerfile for it to achieve our goals
2. Changes to INSTALLATION.md so that people know what to do.

There is another issue to test whether the app starts. You'll need to figure out a way for the app starting test and docker app starting test can work together.

palisadoes commented:
During the week of November 11, 2024 we will start a code freeze on the `develop` branches in Talawa, Talawa Admin and Talawa-API. 

We have completed a project to convert the Talawa-API backend to use PostgreSQL. Work will then begin with us merging code in the `develop` branches to a new `develop-postrgres` branch in each repository.

Planning activities for this will be managed in our #talawa-projects slack channel. A GitHub project will be created to track specially labeled issues. We completed a similar exercise last year using a similar methodology. 

Starting November 12, California time no new PRs will be accepted against the `develop` branch. They must be applied to the `develop-postrgres` branch. 

There are some GSoC project features that will need to be merged into develop. These will be the only exceptions.

This activity and the post GSoC 2024 start date was announced in our #general Slack channel last month as a pinned post.

VanshikaSabharwal commented:
> @VanshikaSabharwal
> 
> You'll need to update this PR with:
> 
> 1. The Dockerfile for it to achieve our goals
> 2. Changes to INSTALLATION.md so that people know what to do.
> 
> There is another issue to test whether the app starts. You'll need to figure out a way for the app starting test and docker app starting test can work together.

Yes, @palisadoes 

Thank you for the feedback!

1. I'll add the Dockerfile to ensure we meet our goals.
2. I'll update INSTALLATION.md with the necessary steps for Docker setup so users can follow along easily.
3. For the app startup tests, I'll work on a solution that allows both the app starting test and the Docker app starting test to function together without conflicts.

Please let me know if there are any additional requirements. Thanks!

VanshikaSabharwal commented:
> During the week of November 11, 2024 we will start a code freeze on the `develop` branches in Talawa, Talawa Admin and Talawa-API.
> 
> We have completed a project to convert the Talawa-API backend to use PostgreSQL. Work will then begin with us merging code in the `develop` branches to a new `develop-postrgres` branch in each repository.
> 
> Planning activities for this will be managed in our #talawa-projects slack channel. A GitHub project will be created to track specially labeled issues. We completed a similar exercise last year using a similar methodology.
> 
> Starting November 12, California time no new PRs will be accepted against the `develop` branch. They must be applied to the `develop-postrgres` branch.
> 
> There are some GSoC project features that will need to be merged into develop. These will be the only exceptions.
> 
> This activity and the post GSoC 2024 start date was announced in our #general Slack channel last month as a pinned post.

Thank you for the update!

Understood regarding the code freeze and the branch updates. I’ll make sure to redirect any new PRs to the develop-postgres branch starting November 12. For any GSoC-related features needing to go into develop, I’ll follow up as needed to confirm the specifics.

I’ll also keep an eye on #talawa-projects in Slack and the GitHub project board to stay aligned with the planning and tracking.

VanshikaSabharwal commented on docker-compose.yaml:
```diff
@@ -0,0 +1,11 @@
+services:
+  app:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    environment:
+      - REACT_APP_TALAWA_URL=${REACT_APP_TALAWA_URL}
+    # volumes:
+    # - .:/usr/src/app
+    ports:
+      - 4321:4321
```
Did it. 
![image](https://github.com/user-attachments/assets/8fe49efe-8e21-4594-adea-eedc581bfcb6)

VanshikaSabharwal commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+WORKDIR /usr/src/app
+
+COPY . .
+
```
did it.

VanshikaSabharwal commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+COPY . .
+
+RUN npm install -g typescript
+
+RUN npm install
+
```
Did it. 
![image](https://github.com/user-attachments/assets/cde8417c-ad36-405c-ac95-dc7fb2299378)

VanshikaSabharwal commented on docker-compose.yaml:
```diff
@@ -0,0 +1,21 @@
+    build:
+      context: .
+      dockerfile: Dockerfile
+    environment:
+      - REACT_APP_TALAWA_URL=${REACT_APP_TALAWA_URL}
+    ports:
+      - 4321:4321
```
did it. 
![image](https://github.com/user-attachments/assets/838b364f-767d-4bd7-9f45-6147a808b71f)

VanshikaSabharwal commented:
I have merged https://github.com/PalisadoesFoundation/talawa-admin/pull/2418#event-15244530247 into this PR as suggested by @palisadoes.

VanshikaSabharwal commented on docker-compose.yaml:
```diff
@@ -0,0 +1,22 @@
+    ports:
+      - 4321:4321
+    user: node
+    healthcheck:
+      test: ['CMD', 'curl', '-f', 'http://localhost:4321']
+      interval: 30s
+      timeout: 10s
+      retries: 5
+      start_period: 40s
```
did it. 
![image](https://github.com/user-attachments/assets/3166955c-4227-4cd5-a8fe-97fc237f5069)

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -253,6 +253,65 @@ jobs:
+          docker run -d --name talawa-admin-app-container -p 4321:4321 talawa-admin-app
+          echo "Docker container started successfully"
+
+      - name: Check if Talawa Admin App is running
+        run: |
+          timeout="${HEALTH_CHECK_TIMEOUT:-120}"
+          echo "Starting health check with ${timeout}s timeout"
+          while ! nc -z localhost 4321 && [ $timeout -gt 0 ]; do
+            sleep 1
+
+            timeout=$((timeout-1))
+            if [ $((timeout % 10)) -eq 0]; then
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            exit 1
+          fi
+          echo "Port check passed, verifying health endpoint..."
+          curl --fail --silent http://localhost:4321/health || exit 1
+          echo "Health check passed successfully"
```
did it. 
![image](https://github.com/user-attachments/assets/ed7cf64e-6ba2-4f8c-9354-9cf02aa03ed1)

VanshikaSabharwal commented:
Now this PR also solves https://github.com/PalisadoesFoundation/talawa-admin/issues/1075 issue by adding Dockerfile,.docker-ignore and docker-compose file.

palisadoes commented:
1. Remember the INSTALLATION.md file
2. Why is the docker check failing?

VanshikaSabharwal commented:
> 1. Remember the INSTALLATION.md file
> 2. Why is the docker check failing?

Yes, I will update INSTALLATION.md and also i think that there is no health check point due to which the docker test is failing.

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,76 @@
+          echo "Docker container started successfully"
+
+      - name: Check if Talawa Admin App is running
+        run: |
+          timeout="${HEALTH_CHECK_TIMEOUT:-120}"
+          echo "Starting health check with ${timeout}s timeout"
+          while ! nc -z localhost 4321 && [ $timeout -gt 0 ]; do
+            sleep 1
+
+            timeout=$((timeout-1))
+            if [ $((timeout % 10)) -eq 0 ]; then
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            echo "Container logs:"
+            docker logs talawa-admin-app-container
+            exit 1
+          fi
```
there is no health check point in app

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,78 @@ jobs:
+          echo "Container logs:"
+          docker logs talawa-admin-app-container
+          exit 1
+        fi
```
did it

VanshikaSabharwal commented:
@palisadoes I have updated the INSTALLATION.md file and deleted the docker-compose.yml file, as I thought there was no specific reason to keep it, and it was causing issues.

VanshikaSabharwal commented on INSTALLATION.md:
````diff
@@ -145,6 +148,28 @@ npm install
+```
+
+This will expose the container's port 4321 on your local machine, making the application accessible at http://localhost:4321.
+
````
did it

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,76 @@ jobs:
       - name: Validate Documents
         run: graphql-inspector validate './src/GraphQl/**/*.ts' './talawa-api/schema.graphql'
 
+  Docker-Start-Check:
```
Please ensure that this does not run if the PR is submitted by the dependabot user

VanshikaSabharwal replied:
Yes, I did it. 
![image](https://github.com/user-attachments/assets/b90a2e1a-01ce-4a59-ab17-74483b6d6104)

palisadoes reviewed (changes requested):
Please see requested changes

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,77 @@ jobs:
+      - name: Checkout the Repository
+        uses: actions/checkout@v4
+
+      - name: Set up Docker
+        uses: docker/setup-buildx-action@v3
+        with:
+          driver-opts: |
+            image=moby/buildkit:latest
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
```
I removed as suggested. 
![image](https://github.com/user-attachments/assets/6462d2d4-a34a-4725-a7f9-1323f99ddee1)

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,77 @@
+          docker run -d --name talawa-admin-app-container -p 4321:4321 talawa-admin-app
+          echo "Docker container started successfully"
+
+      - name: Check if Talawa Admin App is running
+        run: |
+          timeout="${HEALTH_CHECK_TIMEOUT:-120}"
+          echo "Starting health check with ${timeout}s timeout"
+          while ! nc -z localhost 4321 && [ $timeout -gt 0 ]; do
+            sleep 1
+
+            timeout=$((timeout-1))
+            if [ $((timeout % 10)) -eq 0 ]; then
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            echo "Container logs:"
+            docker logs talawa-admin-app-container
+            exit 1
+          fi
+          echo "Port check passed, verifying health endpoint..."
+
```
+         echo "Port check passed, verifying application..."
+         # Wait for the application to fully initialize
+         sleep 5
+         # Verify the application is responding
+         if ! curl --fail --silent --max-time 10 "http://localhost:4321/"; then
+           echo "Application health check failed"
+           docker logs talawa-admin-app-container
+           exit 1
+         fi
+         echo "Application is healthy". these lines were causing issue as we do not have a health check point and the workflow was failing even when the talawa-app was running in docker

VanshikaSabharwal commented on INSTALLATION.md:
```diff
@@ -145,6 +148,38 @@ npm install
 
+# Installation using Docker
+
+## Prerequisites
+
+1. Install Docker on your system:
+   - [Docker Desktop for Windows/Mac](https://www.docker.com/products/docker-desktop)
+   - [Docker Engine for Linux](https://docs.docker.com/engine/install/)
+
```
I removed the docker compose as it was not needed

VanshikaSabharwal commented on .github/workflows/pull-request.yml:
```diff
@@ -240,19 +240,75 @@
+          docker run -d --name talawa-admin-app-container -p 4321:4321 talawa-admin-app
+          echo "Docker container started successfully"
+
+      - name: Check if Talawa Admin App is running
+        run: |
+          timeout="${HEALTH_CHECK_TIMEOUT:-120}"
+          echo "Starting health check with ${timeout}s timeout"
+          while ! nc -z localhost 4321 && [ $timeout -gt 0 ]; do
+            sleep 1
+
+            timeout=$((timeout-1))
+            if [ $((timeout % 10)) -eq 0 ]; then
+              echo "Still waiting for app to start... ${timeout}s remaining"
+            fi
+          done
+
+          if [ $timeout -eq 0 ]; then
+            echo "Timeout waiting for application to start"
+            echo "Container logs:"
+            docker logs talawa-admin-app-container
+            exit 1
+          fi
+          echo "Port check passed, verifying health endpoint..."
+
```
I did it. 
![image](https://github.com/user-attachments/assets/a297516e-1c02-4503-a70b-e9a96bbbecfe)

VanshikaSabharwal commented on INSTALLATION.md:
`````diff
@@ -145,24 +129,58 @@ npm install
 
 The prerequisites are now installed. The next step will be to get the app up and running.
 
+# Installation using Docker
+
+## Prerequisites
+
+1. Install Docker on your system:
+   - [Docker Desktop for Windows/Mac](https://www.docker.com/products/docker-desktop)
+   - [Docker Engine for Linux](https://docs.docker.com/engine/install/)
+
+### Development Setup
+
+If you prefer to use Docker, you can install the app using the following command:
+
+1.  Create a `.env` file as described in the Configuration section
+
+2.  Build the Docker Image:
+
+Run the following command to build the Docker image:
+
+````bash
+docker build -t talawa-admin .
+```bash
+
+3.  Run the Docker container:
+
+After the build is complete, run the Docker container using this command:
+
+```bash
+docker run -p 4321:4321 talawa-admin
+```bash
+
+The application will be accessible at `http://localhost:4321`
+
`````
did it. 
![image](https://github.com/user-attachments/assets/e5ea5551-f2aa-47ab-aea1-02c44604e025)

VanshikaSabharwal commented:
This is video demonstrating that everything is working well
https://github.com/user-attachments/assets/83498abd-e3b7-4dff-b4d8-c61cfda522dd

VanshikaSabharwal commented:
I have video of building docker image as well but video is too long

palisadoes commented:
@VanshikaSabharwal Please open another PR with the same changes against the `develop-postgres` branch

VanshikaSabharwal commented:
> @VanshikaSabharwal Please open another PR with the same changes against the `develop-postgres` branch

okay
